core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1177,6 +1177,8 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       validBytesCount += batchSize
 
       val batchCompression = CompressionType.forId(batch.compressionType.id)
+      // V2: only one batch regardless of compression
```

Comment:
This is not very accurate. `analyzeAndValidateRecords()` is called in the follower append too, which could include more than one batch. It's just that sourceCompression is only used on the leader path, which only contains one batch currently.